otter/metrics.py

```diff
@@ -44,54 +50,114 @@
     'deleting FROM scaling_group WHERE "tenantId" IN ({tids})')
 
 
-@defer.inlineCallbacks
+def valid_group_row(row):
+    return (
+        row.get('created_at') is not None and
+        row.get('desired') is not None and
+        row.get('status') not in ('DISABLED', 'ERROR') and
+        not row.get('deleting', False))
+
+
 def get_specific_scaling_groups(client, tenant_ids):
     tids = ', '.join("'{}'".format(tid) for tid in tenant_ids)
     query = QUERY_GROUPS_OF_TENANTS.format(tids=tids)
-    results = yield client.execute(query, {}, ConsistencyLevel.ONE)
-    defer.returnValue(r for r in results
-                      if r.get('created_at') is not None and
-                      r.get('desired') is not None and
-                      r.get('status') not in ('DISABLED', 'ERROR') and
-                      not r.get('deleting', False))
+    d = client.execute(query, {}, ConsistencyLevel.ONE)
+    return d.addCallback(filter(valid_group_row))
+
+
+def get_last_info(fname):
+    eff = Effect(ReadFileLines(fname)).on(
+        lambda lines: (int(lines[0]),
+                       datetime.utcfromtimestamp(float(lines[1]))))
+
+    def log_and_return(e):
+        _eff = err(e, "error reading last tenant")
+        return _eff.on(lambda _: (None, None))
+
+    return eff.on(error=log_and_return)
+
+
+def update_last_info(fname, tenants_len, time):
+    eff = Effect(
+        WriteFileLines(
+            fname, [tenants_len, datetime_to_epoch(time)]))
+    return eff.on(error=lambda e: err(e, "error updating last tenant"))
+
+
+@attr.s
+class GetAllGroups(object):
+    pass
+
+
+def get_todays_tenants(tenants, today, last_tenants_len, last_date):
+    """
+    Get tenants that are enabled till today
+    """
+    batch_size = 5
+    tenants = sorted(tenants)
+    if last_tenants_len is None:
+        return tenants[:batch_size], batch_size, today
+    days = (today - last_date).days
+    if days <= 0:
+        return tenants[:last_tenants_len], last_tenants_len, today
+    if len(tenants) < last_tenants_len + batch_size:
```

Comment:
Right, they just seem to be equivalent.  Since if there are fewer tenants than the `last_tenants_len + batch_size`, you just return the tenants that are there, which `tenants = tenants[:last_tenants_len + batch_size]` would also return:

```
>>> [1, 2, 3][:50]
[1, 2, 3]
```

And the length is `len(tenants)`.

If there are more than `last_tenants_len + batch_size` tenants, then `tenants = tenants[:last_tenants_len + batch_size]` would truncate them, and `len(tenants) == last_tenants_len + batch_size`, so the logic seems to work for both cases.